File: python/structurify/resources/templates.py
"""
Templates Resource

Copyright (c) 2026 REDSCVRY TECHNOLOGY PRIVATE LIMITED
Licensed under the MIT License.
"""

from typing import TYPE_CHECKING, List, Dict, Any

if TYPE_CHECKING:
    from structurify.client import Structurify


class TemplatesResource:
    """
    Resource for managing project and column templates.

    Templates define the extraction schema for projects.
    """

    def __init__(self, client: "Structurify"):
        self._client = client

    def list(self) -> List[Dict[str, Any]]:
        """
        List all available project templates.

        Returns:
            List of project templates with their column definitions.

        Example:
            templates = client.templates.list()
            for template in templates:
                print(template["name"], template["id"])
        """
        response = self._client.get("/project-templates")
        return response.get("templates", [])

    def list_columns(self) -> List[Dict[str, Any]]:
        """
        List all available column templates.

        Returns:
            List of column templates.

        Example:
            columns = client.templates.list_columns()
            for column in columns:
                print(column["label"], column["format"])
        """
        response = self._client.get("/templates")
        return response.get("templates", [])

    def get(self, template_id: str) -> Dict[str, Any]:
        """
        Get a specific project template by ID.

        Args:
            template_id: The template ID (e.g., "tpl_invoice")

        Returns:
            Template details with columns.

        Raises:
            NotFoundError: If template not found
        """
        response = self._client.get(f"/project-templates/{template_id}")
        return response.get("template", response)

File: nodejs/src/resources/templates.ts
/**
 * Templates Resource
 *
 * Copyright (c) 2026 REDSCVRY TECHNOLOGY PRIVATE LIMITED
 * Licensed under the MIT License.
 */

import type { Structurify } from '../client';
import type { ProjectTemplate, ColumnTemplate } from '../types';

interface ListTemplatesResponse {
  success: boolean;
  templates: ProjectTemplate[];
}

interface ListColumnTemplatesResponse {
  success: boolean;
  templates: ColumnTemplate[];
}

interface GetTemplateResponse {
  success: boolean;
  template: ProjectTemplate;
}

export class TemplatesResource {
  private client: Structurify;

  constructor(client: Structurify) {
    this.client = client;
  }

  /**
   * List all available project templates.
   */
  async list(): Promise<ProjectTemplate[]> {
    const response = await this.client.get<ListTemplatesResponse>('/project-templates');
    return response.templates ?? [];
  }

  /**
   * List all available column templates.
   */
  async listColumns(): Promise<ColumnTemplate[]> {
    const response = await this.client.get<ListColumnTemplatesResponse>('/templates');
    return response.templates ?? [];
  }

  /**
   * Get a specific project template by ID.
   */
  async get(templateId: string): Promise<ProjectTemplate> {
    const response = await this.client.get<GetTemplateResponse>(
      `/project-templates/${templateId}`
    );
    return response.template;
  }
}
